### PR TITLE
Graduate Evented PLEG to Beta

### PR DESCRIPTION
**Pull request by harche**
Signed-off-by: Harshal Patil [harpatil@redhat.com](mailto:harpatil@redhat.com)

Add EventedPLEG to FeatureGates doc.

Corresponding PR - https://github.com/kubernetes/kubernetes/pull/115967
Enhancement - https://github.com/kubernetes/enhancements/issues/3386

### CODE REVIEW

File: content/en/docs/reference/command-line-tools-reference/feature-gates.md [hunk 1]
```diff
@@ -94,7 +94,7 @@ For a reference to old feature gates that are removed, please refer to
 | `DownwardAPIHugePages` | `false` | Beta | 1.21 | 1.21 |
 | `DownwardAPIHugePages` | `true` | Beta | 1.22 | |
 | `DynamicResourceAllocation` | `false` | Alpha | 1.26 | |
-| `EventedPLEG` | `false` | Alpha | 1.26 | - |
```
Comment (tengqm) [inline content/en/docs/reference/command-line-tools-reference/feature-gates.md:97]: You are not supposed to remove the old line 97. It was and still is about the Alpha stage.
You can revise it to indicate that Alpha ends at 1.26 as well.

**after** this revision, you can add a new line for Beta stage as you'd already done.

File: content/en/docs/reference/command-line-tools-reference/feature-gates.md [hunk 2]
```diff
@@ -94,7 +94,7 @@ For a reference to old feature gates that are removed, please refer to
...
+| `EventedPLEG` | `false` | Beta | 1.27 | - |
```
Comment (sftim) [inline content/en/docs/reference/command-line-tools-reference/feature-gates.md:97]:
```suggestion
| `EventedPLEG` | `false` | Alpha | 1.26 | 1.26 |
| `EventedPLEG` | `false` | Beta | 1.27 | - |
```
Try that?

**Review by sftim — COMMENTED** [review]
Thanks @harche

(I think that) you should also update https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/ to mention this beta feature gate and why you might care about it.

If that's not the right page to update, we should find which is the right page and update that instead. For beta features, even those disabled by default, we nearly document why you might want to turn them on and the work you need to do associated with enabling the feature.

If you think there are tasks to document - eg: _migrate to using CRI notifications for Pod events_ - then please add that task page either in this PR or in a separate PR that also mentions the relevant KEP.

File: content/en/docs/reference/command-line-tools-reference/feature-gates.md (hunk 2 above)
Comment (harche) [reply content/en/docs/reference/command-line-tools-reference/feature-gates.md:98]: Thanks for that pointer.

**Comment by harche** [comment]
> Thanks @harche
> 
> (I think that) you should also update https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/ to mention this beta feature gate and why you might care about it.
> 
> If that's not the right page to update, we should find which is the right page and update that instead. For beta features, even those disabled by default, we nearly document why you might want to turn them on and the work you need to do associated with enabling the feature.
> 

IMO, the right place for such page would https://kubernetes.io/docs/concepts/configuration/ 

I will add a page describing Evented PLEG and its usage. Thanks.

**Comment by sftim** [comment]
> IMO, the right place for such page would https://kubernetes.io/docs/concepts/configuration/

That _isn't_ the best home. If you want to document pod lifecycle events as a concept, put that in https://kubernetes.io/docs/concepts/containers/ or https://kubernetes.io/docs/concepts/workloads/pods/

If you want to explain the cluster operator angle, the content should live in either https://kubernetes.io/docs/concepts/cluster-administration/ or https://kubernetes.io/docs/tasks/administer-cluster/ (make the choice depending on whether you are explaining how to perform a task, such as migrating to CRI notifications, or providing general context).

**Comment by harche** [comment]
> > IMO, the right place for such page would https://kubernetes.io/docs/concepts/configuration/
> 
> That _isn't_ the best home. If you want to document pod lifecycle events as a concept, put that in https://kubernetes.io/docs/concepts/containers/ or https://kubernetes.io/docs/concepts/workloads/pods/
> 
> If you want to explain the cluster operator angle, the content should live in either https://kubernetes.io/docs/concepts/cluster-administration/ or https://kubernetes.io/docs/tasks/administer-cluster/ (make the choice depending on whether you are explaining how to perform a task, such as migrating to CRI notifications, or providing general context).

It's a configuration of the kubelet (and CRI Runtime) that affects how it maintains its internal cache of pod statuses. That's the reason I thought configuration made sense. This is not directly connected to any concepts related to containers or pods, neither is it related to any operator or cluster administration.

**Comment by harche** [comment]
Coming to think of it, maybe https://kubernetes.io/docs/tasks/administer-cluster/ makes sense.

File: content/en/docs/tasks/administer-cluster/switch-to-evented-pleg.md
Comment (sftim) [inline content/en/docs/tasks/administer-cluster/switch-to-evented-pleg.md:1]: Let's not call this “evented PLEG” as this is jargon many cluster operators won't understand.

Instead, talk about switching from polling to event-based updates to container status.

File: content/en/docs/tasks/administer-cluster/switch-to-evented-pleg.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+
+
+<!-- steps -->
+
+## Why switch to Evented PLEG?
+
+* The current `Generic PLEG` incurs non-negligible overhead due to frequent polling of container statuses.
+* This overhead is exacerbated by Kubelet's parallelism, limiting its scalability and causing poor performance and reliability problems.
+* The goal of `Evented PLEG` is to reduce unnecessary work during inactivity by replacing periodic polling.
```
Comment (sftim) [inline content/en/docs/tasks/administer-cluster/switch-to-evented-pleg.md:24]: This should go before the list of prerequisites.

File: content/en/docs/tasks/administer-cluster/switch-to-evented-pleg.md [hunk 4]
```diff
@@ -0,0 +1,59 @@
+---
+title: Switch to Evented PLEG
+content_type: task
+weight: 90
+---
```
Comment (sftim) [inline content/en/docs/tasks/administer-cluster/switch-to-evented-pleg.md:5]:
```suggestion
min-kubernetes-server-version: 1.26
---
```

**Review by sftim — COMMENTED** [review]
Thanks. This will need some work before we can merge it, I'm afraid. For example, I think we should advice readers to drain the nodes they're reconfiguring, or at least empower them to consider doing that.

What do you think @harche ?

**Comment by harche** [comment]
> Thanks. This will need some work before we can merge it, I'm afraid. For example, I think we should advice readers to drain the nodes they're reconfiguring, or at least empower them to consider doing that.
> 
> What do you think @harche ?

yes, it makes sense.

File: content/en/docs/tasks/administer-cluster/switch-to-evented-pleg.md [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+
+<!-- steps -->
+
+## Why switch to Evented PLEG?
+
+* The current `Generic PLEG` incurs non-negligible overhead due to frequent polling of container statuses.
+* This overhead is exacerbated by Kubelet's parallelism, limiting its scalability and causing poor performance and reliability problems.
+* The goal of `Evented PLEG` is to reduce unnecessary work during inactivity by replacing periodic polling.
+
```
[the remark this answers is not in the document]
Comment (harche) [reply content/en/docs/tasks/administer-cluster/switch-to-evented-pleg.md:43]: yes.

File: content/en/docs/tasks/administer-cluster/switch-to-evented-pleg.md (hunk 4 above)
Comment (harche) [reply content/en/docs/tasks/administer-cluster/switch-to-evented-pleg.md:6]: Updated. Thanks.

File: content/en/docs/tasks/administer-cluster/switch-to-evented-pleg.md [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+<!-- overview -->
+This page shows how to use `Evented PLEG` in Kubelet which attempts to lower the resource consumption by avoiding frequent polling for container statuses.
+
+
+## {{% heading "prerequisites" %}}
+
+
+{{< include "task-tutorial-prereqs.md" >}} {{< version-check >}}
+
```
[the remark this answers is not in the document]
Comment (harche) [reply content/en/docs/tasks/administer-cluster/switch-to-evented-pleg.md:32]: > * Do you need to enable any feature gates on the control plane?
No.

> * Do you need a specific Kubernetes version?

1.26+

> * Do you need administrator-level access to the node(s) you're managing?

Yes

> * Do you need a container runtime that has support for a particular CRI feature?

Yes

> 
> and we should also cover, somewhere in this page:
> 
> * do you need to roll this out to all nodes at once, or can you phase it?

You can phase it. It is only a node specific feature that affects how the given kubelet manages the pod status cache internally.

File: content/en/docs/tasks/administer-cluster/switch-to-evented-pleg.md [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+## Switching to Evented PLEG
+
+1. Start the `CRI Runtime` with the `Evented PLEG` support. e.g.
+   1. Containerd version 1.7+ 
+   2. CRI-O v1.26+ and also, start the cri-o daemon with the flag `--enable-pod-events=true` or using a drop in config like,
+      ```toml
+      [crio.runtime]
+      enable_pod_events: true
+      ```
```
[the remark this answers is not in the document]
Comment (harche) [reply content/en/docs/tasks/administer-cluster/switch-to-evented-pleg.md]: Updated. Thanks.

File: content/en/docs/tasks/administer-cluster/switch-to-evented-pleg.md
Comment (harche) [reply content/en/docs/tasks/administer-cluster/switch-to-evented-pleg.md:1]: Updated the Title.

File: content/en/docs/tasks/administer-cluster/switch-to-evented-pleg.md [hunk 8]
```diff
@@ -0,0 +1,76 @@
+---
+title: Switching from polling to CRI event-based updates to container status
```
Comment (sftim) [inline content/en/docs/tasks/administer-cluster/switch-to-evented-pleg.md:2]:
```suggestion
title: Switching From Polling to CRI Event-based Updates to Container Status
```

**Comment by harche** [comment]
@sftim Thanks for your valuable feedback. I have updated the PR with those recommendations.

File: content/en/docs/tasks/administer-cluster/switch-to-evented-pleg.md [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+min-kubernetes-server-version: 1.26
+content_type: task
+weight: 90
+---
+
+{{< feature-state for_k8s_version="v1.26" state="beta" >}}
+
+<!-- overview -->
+This page shows how to use `Evented PLEG` (CRI event-based updates to container status) in Kubelet which attempts to lower the resource consumption by avoiding frequent polling for container statuses.
```
Comment (sftim) [inline content/en/docs/tasks/administer-cluster/switch-to-evented-pleg.md:11]:
```suggestion
This page shows how to migrate notes to use event based updates for container status. The event-based
implementation reduces node resource consumption by the kubelet, compared to the legacy approach
that relies on polling.

You may know this feature as _evented Pod lifecycle event generator (PLEG)_. That's the name used
internally within the Kubernetes project for a key implementation detail.
```
?

File: content/en/docs/tasks/administer-cluster/switch-to-evented-pleg.md [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+
+
+## {{% heading "prerequisites" %}}
+
+* This support for this feature is available in Kubernetes {{< skew currentVersion >}};
+  Kubernetes v1.26 as a beta feature, disabled by default.
+* Please make sure the node is [drained](https://kubernetes.io/docs/tasks/administer-cluster/safely-drain-node/) before proceeding.
```
Comment (sftim) [inline content/en/docs/tasks/administer-cluster/switch-to-evented-pleg.md:18]: This should be part of the steps and not a prerequisite.

File: content/en/docs/tasks/administer-cluster/switch-to-evented-pleg.md [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+      Version 1.7+
+      {{< /tab >}}
+      {{< tab name="CRI-O" codelang="bash" >}}
+      v1.26+ and also, start the cri-o daemon with the flag `--enable-pod-events=true` or using a drop in config like,
+            
+      [crio.runtime]
+      enable_pod_events: true
+      
+      {{< /tab >}}
+      {{< /tabs >}}
```
Comment (sftim) [inline content/en/docs/tasks/administer-cluster/switch-to-evented-pleg.md:31]: At this point, we should have the reader check that their runtime _can_ support events. They don't have to have already reconfigured it (eg CRI-O and `enable_pod_events`).

**Review by sftim — COMMENTED** [review]
Some more feedback

**Pull request merged**